pyflowater/flostream.py

```diff
@@ -0,0 +1,56 @@
+"""Code relating to the firestore API Flo use to stream live data."""
+
+import requests
+import json
+from google.cloud import firestore
+from google.oauth2.credentials import Credentials
+
+from pyflowater.const import (
+    FLO_GOOGLE_API_KEY,
+    FIREBASE_REST_API,
+    FLO_FIRESTORE_PROJECT,
+)
+
+def _get_token_info(token):
+    url = f"{FIREBASE_REST_API}/verifyCustomToken?key={FLO_GOOGLE_API_KEY}"
+    headers = {"Content-type": "application/json; charset=UTF-8"}
+    data = json.dumps({"returnSecureToken": True, "token": token})
+    resp = requests.post(url, headers=headers, data=data)
+    return resp.json()
+
+class FloListener:
+    """Flo firestore listener class."""
+
+    def __init__(self, token, deviceId, callback):
+        self._token = token
+        self._deviceId = deviceId
+        self._callback = callback
+        self._watch = None
+        self._client = None
+        self._doc_ref = None
+
+    def set_callback(self, callback):
+        """Update the callback."""
+        self._callback = callback
+
+    def start(self):
+        """Begin listening on the firestore database."""
+        if self._watch:
+            return
+        if not self._client:
+            tinfo = _get_token_info(self._token)
+            creds = Credentials(tinfo['idToken'], refresh_token=tinfo['refreshToken'])
+            self._client = firestore.Client(project=FLO_FIRESTORE_PROJECT, credentials=creds)
+        if not self._doc_ref:
+            self._doc_ref = self._client.collection('devices').document(self._deviceId)
+        self._watch = self._doc_ref.on_snapshot(self._handle)
+
+    def stop(self):
+        """Shut down the listener, if started."""
+        if not self._watch:
+            return
+        self._watch.close()
+        self._watch = None
+
+    def _handle(self, document, changes, timestamp):
+        self._callback(document[0].to_dict())
```

Comment:
Since server costs are per-document returned to the listener, it would probably be good to emit a warming if len(document) > 1. Shouldn't ever happen but if it does the warning should ask them to raise an issue here.

Reply:
Indeed, but the documentation also mentions that since each callback on the listener includes the updated document, the cost will be incurred for each callback (which appears to be ~twice a second), so the service costs are going to add up while using the streaming function anyway. It does explain why they require the server to be pinged with the heartbeat (which I forgot to include in this patch - if the heartbeat stops, they also stop exporting readings into firestore).

Reply:
I see. This is very interesting. I had thought that they'd only update documents when the values change, which is not very often. But they update twice per second even if nothing is changing? Huh, not the choice I'd have made. So please confirm that they update the document even if doesn't change.

Reply:
Yes - I'm not sure exactly why, but it has a fairly large blob of JSON. The order of the content in the JSON changes, so it's like it updates on the fly, newest first (or last). Here's an example (redacting my serial number):
```
{
    "valve": {
        "lastKnown": "open"
    },
    "telemetry": {
        "current": {
            "psi": 69.2,
            "updated": "2021-09-04T23:07:03Z",
            "gpm": 0.0,
            "tempF": 97.0
        }
    },
    "systemMode": {
        "lastKnown": "home"
    },
    "waterConsumption": {
        "estimateToday": 48.502,
        "estimateLastUpdated": "2021-09-04T23:07:02Z"
    },
    "_latestWrite": "2021-09-04T23:07:03Z",
    "connectivity": {
        "rssi": -46.0
    },
    "healthTest": {
        "roundId": "<redacted-GUID-with-device-ID>",
        "updated": "2021-09-04T10:31:02Z",
        "fsUpdated": "2021-09-04T23:05:39Z",
        "status": "completed",
        "type": "auto"
    },
    "value": {
        "deviceId": "<redacted>"
    },
    "floDetect": {
        "latestEvent": {
            "incident": {
                "id": "00000000-0000-0000-0000-000000000000"
            },
            "startAt": "2021-09-04T23:04:24Z",
            "predicted": {
                "id": 300.0
            },
            "totalGal": 0.509,
            "id": "<redacted-random-GUID>",
            "endAt": "2021-09-04T23:05:01Z"
        }
    },
    "notifications": {
        "fsUpdate": "2021-09-04T10:31:05Z",
        "pending": {
            "alarmCount": [],
            "criticalCount": 0.0,
            "warningCount": 0.0,
            "infoCount": 0.0
        }
    },
    "deviceId": "<redacted>",
    "isConnected": true,
    "shutoff": {
        "scheduledAt": "1970-01-01T00:00:00Z",
        "alertId": ""
    }
}
```
I was mainly interested in the telemetry dictionary for Prometheus, so I wrote a script to grab that, but I did print out the `updated` key to check it was working:
```
Updated at 2021-09-04T23:56:57Z
Updated at 2021-09-04T23:56:59Z
Updated at 2021-09-04T23:56:59Z
Updated at 2021-09-04T23:56:59Z
Updated at 2021-09-04T23:56:59Z
Updated at 2021-09-04T23:58:09Z
Updated at 2021-09-04T23:58:09Z
Updated at 2021-09-04T23:58:11Z
Updated at 2021-09-04T23:58:13Z
Updated at 2021-09-04T23:58:13Z
Updated at 2021-09-04T23:58:15Z
Updated at 2021-09-04T23:58:15Z
Updated at 2021-09-04T23:58:17Z
Updated at 2021-09-04T23:58:17Z
Updated at 2021-09-04T23:58:19Z
Updated at 2021-09-04T23:58:19Z
```
Sadly I didn't print out walltime simultaneously, but as you can see, it does sometimes take a break, but there's still regular updates, and it looks to even out to about once per second, even though the data isn't updating that fast. I can also see that on at least two consecutive updates, the only value that changed was the `updated` value within the telemetry section (e.g. the actual `gpm`/`psi`/`tempF` values stayed the same), and it also caused the outermost `_latestWrite` value to change, too.

I might leave it running a little longer to see if it settles down (e.g. at night, when no water is use so the water pressure can stabilise), but since I had been hoping to use streaming constantly I am concerned it might run up a bit of a bill for Moen (my guess is ~$5 a month for this level of activity, also factoring in their write costs). Though it seems like they could probably save some money with firmware tweaks :)